### ISSUE TITLE
Suspicious array indices in device-routines.C

### ISSUE THREAD

**Issue by rwvo** [geodynamics/sw4lite#9]
When compiling sw4lite with a clang-based compiler for AMD GPUs, I'm getting warnings for device-routines.C, lines 1841 and 1843, e.g.:

```
src/device-routines.C:1841:26: warning: array index 3 is past the end of the array (which contains 3 elements)
      [-Warray-bounds]
            (qum[c][4]-2*qum[c][3]+sum(c,ith,jth,k  ))
                         ^      ~
src/device-routines.C:1744:3: note: array 'qum' declared here
  float_sw4 qu[DIAMETER][3], qum[DIAMETER][3];                                                                            ^                      
                                                                        
```
Since DIAMETER has the value 5, and in the loop, c is at most 3, the array accesses are not really out of bounds, but I can't tell whether this is an actual mistake, or if something clever is going on. Could you please double-check, and perhaps add a comment if the code is in fact correct?

The relevant lines were added by Anders Petersson in Oct last year:
```
[rvanoo@snell:~/repos/sw4lite] $ git blame -L 1840,+5 src/device-routines.C                                            
^dfbecf0 (Anders Petersson 2017-10-19 16:07:10 -0700 1840)             -rho(i,j,k+1)*dcz(k+1)*                         
^dfbecf0 (Anders Petersson 2017-10-19 16:07:10 -0700 1841)             (qum[c][4]-2*qum[c][3]+sum(c,ith,jth,k  ))      
^dfbecf0 (Anders Petersson 2017-10-19 16:07:10 -0700 1842)             +2*rho(i,j,k)*dcz(k)  *                         
^dfbecf0 (Anders Petersson 2017-10-19 16:07:10 -0700 1843)             (qum[c][3]-2*sum(c,ith,jth,k  )+qum[c][1])      
^dfbecf0 (Anders Petersson 2017-10-19 16:07:10 -0700 1844)             -rho(i,j,k-1)*dcz(k-1)*                         
```

[Edit] I just noticed that the commit mentioned by 'git blame' above was the initial commit.

**Comment by andersp** [comment]
Somebody from Nvidia wrote that code and we are not sure what the intention was. However, it is clearly a bug. If you declare an array in C with 3 components, you can only access it with [0], [1], or [2]. So, for float_sw4 qu[DIAMETER][3],the second index should only be [0], [1], or [2].
/Anders

**Comment by bjorn2** [comment]
In the latest commit, by Pei-Hung, this bug appears to be fixed. Try git pull and recompile!
/Bjorn

**Comment by rwvo** [comment]
Bjorn: thanks, I cloned the repo earlier this afternoon, and when I just pulled again, I got 10 new commits. I'll have a look.

Anders: this is a two-dimensional array declared as float_sw4 qu[5][3], and accessing qu[2][3] is in this case equivalent to accessing qu[3][0]. Ugly, but I've seen people do this on purpose, typically to simplify code/offset calculations.

Thanks for the quick responses!

**Comment by rwvo** [comment]
Fixed in commit d995bdb9.
